rtrlib/aspa/aspa.c

```diff
@@ -358,17 +446,23 @@ enum aspa_status aspa_table_update(struct aspa_table *aspa_table, struct rtr_soc
 #ifdef ASPA_UPDATE_IN_PLACE
 				   bool revert,
 #endif
-				   struct aspa_update_operation **failed_operation,
+				   struct aspa_update_operation **failed_operation, size_t *failed_index,
 				   struct aspa_update_finalization_args **finalization_args)
 {
 	if (!rtr_socket || !operations || len == 0 || !failed_operation)
 		return ASPA_ERROR;
 
+#ifdef ASPA_UPDATE_IN_PLACE
 	if (!revert) {
+#endif
 		// stable sort operations, so operations dealing with the same customer ASN
 		// are located right next to each other
 		qsort(operations, len, sizeof(struct aspa_update_operation), compare_update_operations);
+#ifdef ASPA_UPDATE_IN_PLACE
+	} else {
+		qsort(operations, len, sizeof(struct aspa_update_operation), compare_update_operations_reverse_index);
 	}
```

Comment:
Why do you need this reverse stuff here?

Reply:
This adds too much complexity.

Reply:
for undoing, the updates need to be processed to in the reverse order in which they arrived (going backwards in time).

e.g. when there is an entry for cas 10,
assume updates are incoming:
1) withdraw 10
2) add 10
3) add 10 (this will cause an error: duplicate add)

after step 2), there is an entry for cas 10.
if we wouldn't process updates backwards in time, we would cause another error: 1') add 10 would cause duplicate add again.

To process updates backwards in time, it is necessary to either
a) reverse the order of operations
b) iterate backwards over operations.

Because b) would cause further changes in existing_i logic, i chose a).

Reply:
Ok I agree on the issue but I'd suggest we modify the existing_i logic (iterate backwards)